lib/galaxy/webapps/galaxy/api/datatypes.py

```diff
@@ -2,14 +2,200 @@
 API operations allowing clients to determine datatype supported by Galaxy.
 """
 import logging
+from typing import (
+    Dict,
+    List,
+    Optional,
+    Union,
+)
 
+from fastapi import (
+    Depends,
+    Query,
+)
+from fastapi.routing import APIRouter
+from fastapi_utils.cbv import cbv
+
+
+from galaxy.app import UniverseApplication
 from galaxy.datatypes.data import Data
+from galaxy.datatypes.registry import Registry
+from galaxy.datatypes.schema import (
+    DatatypeConverter,
+    DatatypeDetails,
+    DatatypesCombinedMap,
+    DatatypesMap,
+)
 from galaxy.util import asbool
 from galaxy.web import expose_api_anonymous_and_sessionless
 from galaxy.webapps.base.controller import BaseAPIController
+from . import get_app
 
 log = logging.getLogger(__name__)
 
+router = APIRouter(tags=['datatypes'])
+
+ExtensionOnlyQueryParam : Optional[bool] = Query(
+    default=True,
+    title="Extension only",
+    description="Whether to return only the datatype's extension rather than the datatype's details",
+)
+
+UploadOnlyQueryParam : Optional[bool] = Query(
+    default=True,
+    title="Upload only",
+    description="Whether to return only datatypes which can be uploaded",
+)
+
+
+def get_datatypes_registry(app: UniverseApplication = Depends(get_app)) -> Registry:
+    return app.datatypes_registry
+
+
+@cbv(router)
+class FastAPIDatatypes:
+    datatypes_registry: Registry = Depends(get_datatypes_registry)
+
+    @router.get('/api/datatypes',
+        summary="Lists all available data types",
+        response_model=Union[List[DatatypeDetails], List[str]],
+        response_description="List of data types")
+    async def index(self,
+                    extension_only: Optional[bool] = ExtensionOnlyQueryParam,
+                    upload_only: Optional[bool] = UploadOnlyQueryParam
+                    ) -> Union[List[DatatypeDetails], List[str]]:
+        """Gets the list of all available data types."""
+        try:
+            return self._index(extension_only, upload_only)
+        except Exception as e:
```

Comment:
Can we do this without catching and rethrowing the exception? It is a lot of extra untested code - if we want all these exceptions logged like this we can do it at the framework level (https://github.com/galaxyproject/galaxy/pull/10823/files#diff-ef136e8d148bc95da2725c2cc8e228e60cffeeb0b84820f7152646d11ced9ff5R11).

Reply:
I see - sometimes the legacy controllers do this also - it is bad when they do it also. Can you remove the exception handling down there also?

Reply:
Sure!